File: STABLE/3.0.0/java/src/org/mariuszgromada/math/mxparser/Miscellaneous.java
/*
 * @(#)Miscellaneous.java        3.0.0    2016-05-07
 *
 * You may use this software under the condition of "Simplified BSD License"
 *
 * Copyright 2010-2016 MARIUSZ GROMADA. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY <MARIUSZ GROMADA> ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of MARIUSZ GROMADA.
 *
 * If you have any questions/bugs feel free to contact:
 *
 *     Mariusz Gromada
 *     mariuszgromada.org@gmail.com
 *     http://mathparser.org
 *     http://mathspace.pl
 *     http://janetsudoku.mariuszgromada.org
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://mariuszgromada.github.io/MathParser.org-mXparser
 *     http://mxparser.sourceforge.net
 *     http://bitbucket.org/mariuszgromada/mxparser
 *     http://mxparser.codeplex.com
 *     http://github.com/mariuszgromada/Janet-Sudoku
 *     http://janetsudoku.codeplex.com
 *     http://sourceforge.net/projects/janetsudoku
 *     http://bitbucket.org/mariuszgromada/janet-sudoku
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *
 *                              Asked if he believes in one God, a mathematician answered:
 *                              "Yes, up to isomorphism."
 */
package org.mariuszgromada.math.mxparser;

import java.util.ArrayList;
import java.util.Comparator;

import org.mariuszgromada.math.mxparser.parsertokens.ConstantValue;
import org.mariuszgromada.math.mxparser.parsertokens.Token;

/*=================================================
*
* Package level classes and interfaces
*
*=================================================
*/
/**
 * Package level class for handling function parameters.
 */
class FunctionParameter {
	ArrayList<Token> tokens;
	String paramStr;
	int fromIndex;
	int toIndex;
	FunctionParameter(ArrayList<Token> tokens,
			String paramStr,
			int fromIndex,
			int toIndex ) {
		this.tokens = tokens;
		this.paramStr = paramStr;
		this.fromIndex = fromIndex;
		this.toIndex = toIndex;
	}
}
/**
* Package level class for generating iterative operator parameters
*/
class IterativeOperatorParameters {
	FunctionParameter indexParam;
	FunctionParameter fromParam;
	FunctionParameter toParam;
	FunctionParameter funParam;
	FunctionParameter deltaParam;
	Expression fromExp;
	Expression toExp;
	Expression funExp;
	Expression deltaExp;
	double from;
	double to;
	double delta;
	boolean withDelta;
	IterativeOperatorParameters(ArrayList<FunctionParameter> functionParameters) {
		/*
		 * Get index string
		 * 1st parameter
		 */
		indexParam = functionParameters.get(0);
		/*
		 * Get from string (range from-to)
		 * 2nd parameter
		 */
		fromParam = functionParameters.get(1);
		/*
		 * Get to string (range from-to)
		 * 3rd parameter
		 */
		toParam = functionParameters.get(2);
		/*
		 * Get internal function strinng
		 * 4th - parameter
		 */
		funParam = functionParameters.get(3);
		/*
		 * Get internal function strinng
		 * 5th - parameter
		 */
		deltaParam = null;
		withDelta = false;
		if (functionParameters.size() == 5) {
			deltaParam = functionParameters.get(4);
			withDelta = true;
		}
	}
}
/**
* Handling argument parameters
*/
class ArgumentParameter {
	Argument argument;
	double initialValue;
	int initialType;
	int presence;
	int index;
	ArgumentParameter() {
		argument = null;
		initialValue = Double.NaN;
		initialType = ConstantValue.NaN;
		presence = Expression.NOT_FOUND;
	}
}
/**
* Base class prepresenting key words knwon by the parsere
*/
class KeyWord {
	String wordString;
	int	wordId;
	int wordTypeId;
	String description;
	public KeyWord() {
		wordString = "";
		wordId = ConstantValue.NaN;
		wordTypeId = ConstantValue.NaN;
		description = "";
	}
	/**
	 * Constructor - creates key words form wordStrin wordId
	 * and wordTypId
	 *
	 * @param wordString   the word string (refers to below interfaces)
	 * @param wordId       the word identifier (refers to below interfaces)
	 * @param wordTypeId   the word type (refers to below interfaces)
	 */
	KeyWord(String wordString, int wordId, int wordTypeId) {
		this.wordString = wordString;
		this.wordId = wordId;
		this.wordTypeId = wordTypeId;
	}
	KeyWord(String wordString, String description, int wordId, int wordTypeId) {
		this.wordString = wordString;
		this.wordId = wordId;
		this.wordTypeId = wordTypeId;
		this.description = description;
	}
}
/**
* Internal token class
* which is used with stack while
* evaluation of tokens levels
*/
class TokenStackElement {
	int tokenIndex;
	int tokenId;
	int tokenTypeId;
	int tokenLevel;
	boolean precedingFunction;
}
class SyntaxStackElement {
	String tokenStr;
	int tokenLevel;
	SyntaxStackElement(String tokenStr, int tokenLevel) {
		this.tokenStr = tokenStr;
		this.tokenLevel = tokenLevel;
	}
}
/*
* ---------------------------------------------------------
* Comparators for sorting
* ---------------------------------------------------------
*/
/**
* Comparator for key word list sorting by key word string.
* This king of sorting is used while checking the syntax
* (duplicated key word error)
*/
class KwStrComparator implements Comparator<KeyWord> {
	/**
	 *
	 */
	public int compare(KeyWord kw1, KeyWord kw2) {
		String s1 = kw1.wordString;
		String s2 = kw2.wordString;
		return s1.compareTo(s2);
	}
}
/**
* Comparator for key word list sorting by
* descending key word length
* .
* This king of sorting is used while tokenizing
* (best match)
*/
class DescKwLenComparator implements Comparator<KeyWord> {
	/**
	 *
	 */
	public int compare(KeyWord kw1, KeyWord kw2) {
		int l1 = kw1.wordString.length();
		int l2 = kw2.wordString.length();
		return l2-l1;
	}
}
/**
* Comparator for key word list sorting by
* type of the key word
*
*/
class KwTypeComparator implements Comparator<KeyWord> {
	/**
	 *
	 */
	public int compare(KeyWord kw1, KeyWord kw2) {
		int t1 = kw1.wordTypeId*1000 + kw1.wordId;
		int t2 = kw2.wordTypeId*1000 + kw2.wordId;
		return t1-t2;
	}
}
/*
* ---------------------------------------------------------
* Grouping constants by interfaces
* ---------------------------------------------------------
*/
/*
* Package level class to be used
* while function, argument, constant definition
* using only one string, ie:
* Function "f(x,y) = sin(x) + cos(y)"
* Constant "a = 5/20"
*/
class HeadEqBody {
	private boolean ONLY_PARSER_KEYWORDS = true;
	String headStr;
	String bodyStr;
	int eqPos;
	ArrayList<Token> headTokens;
	boolean definitionError;
	HeadEqBody(String definitionString) {
		char c;
		eqPos = 0;
		int matchStatus = mXparser.NOT_FOUND;
		definitionError = false;
		do {
			c = definitionString.charAt(eqPos);
			if (c == '=') matchStatus = mXparser.FOUND;
			else eqPos++;
		} while ( (eqPos < definitionString.length()) && (matchStatus == mXparser.NOT_FOUND) );
		if ( (matchStatus == mXparser.FOUND) && (eqPos > 0) && (eqPos <= definitionString.length()-2) ) {
			headStr = definitionString.substring(0, eqPos);
			bodyStr  = definitionString.substring(eqPos+1);
			Expression headExpression = new Expression(headStr, ONLY_PARSER_KEYWORDS);
			headTokens = headExpression.getCopyOfInitialTokens();
		} else {
			definitionError = true;
			headStr = "";
			bodyStr = "";
			headTokens = null;
			eqPos = -1;
		}
	}
}

File: CURRENT/c-sharp/tests-and-release/4-Unit-Tests/Miscellaneous.cs
/*
 * @(#)Miscellaneous.cs        5.0.0    2022-02-12
 *
 * You may use this software under the condition of "Simplified BSD License"
 *
 * Copyright 2010-2022 MARIUSZ GROMADA. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *    1. Redistributions of source code must retain the above copyright notice, this list of
 *       conditions and the following disclaimer.
 *
 *    2. Redistributions in binary form must reproduce the above copyright notice, this list
 *       of conditions and the following disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY <MARIUSZ GROMADA> ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are those of the
 * authors and should not be interpreted as representing official policies, either expressed
 * or implied, of MARIUSZ GROMADA.
 *
 * If you have any questions/bugs feel free to contact:
 *
 *     Mariusz Gromada
 *     mariuszgromada.org@gmail.com
 *     http://mathparser.org
 *     http://mathspace.pl
 *     http://janetsudoku.mariuszgromada.org
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *     http://mariuszgromada.github.io/MathParser.org-mXparser
 *     http://mxparser.sourceforge.net
 *     http://bitbucket.org/mariuszgromada/mxparser
 *     http://mxparser.codeplex.com
 *     http://github.com/mariuszgromada/Janet-Sudoku
 *     http://janetsudoku.codeplex.com
 *     http://sourceforge.net/projects/janetsudoku
 *     http://bitbucket.org/mariuszgromada/janet-sudoku
 *     http://github.com/mariuszgromada/MathParser.org-mXparser
 *
 *                              Asked if he believes in one God, a mathematician answered:
 *                              "Yes, up to isomorphism."
 */

using System;
using org.mariuszgromada.math.mxparser.mathcollection;

namespace org.mariuszgromada.math.mxparser.test {

    /**
     * Common tools for unit tests
     */
    internal class TestCommonTools {
        internal static void consolePrintOptionsMxparser(String prefix) {
            mXparser.consolePrintSettings(prefix);
        }

        internal static void consolePrintOptionsExpression(Expression testExp) {
            mXparser.consolePrintln("------------ OPTIONS AFTER ------------");
            mXparser.consolePrintln("getImpliedMultiplicationMode = " + testExp.getImpliedMultiplicationMode());
            mXparser.consolePrintln("getCanonicalExpressionString = " + testExp.getCanonicalExpressionString());
            mXparser.consolePrintSettings();
            mXparser.consolePrintln("getErrorMessage = " + testExp.getErrorMessage());
            mXparser.consolePrintln("checkLexSyntax = " + testExp.checkLexSyntax());
            mXparser.consolePrintln("checkSyntax = " + testExp.checkSyntax());
            mXparser.consolePrintln("getCopyOfInitialTokens = ");
            mXparser.consolePrintTokens(testExp.getCopyOfInitialTokens());
        }

        internal static void consolePrintTestStart(int testId, String testStr, String testType) {
            mXparser.consolePrintln();
            mXparser.consolePrintln("[" + testType +"/" + testId + "] ------------ TEST INIT: " + testStr);
            mXparser.consolePrintln("-- Options before");
            consolePrintOptionsMxparser("---- ");
            mXparser.consolePrint("[" + testType +"/" + testId + "] Starting test: " + testStr + " ...... ");
        }

        internal static void consolePrintTestExprStart(int testId, String testStr) {
            consolePrintTestStart(testId, testStr, "Expr");
        }

        internal static void consolePrintTestExprEnd(double calculatedValue, double expectedValue, bool testResult, Expression testExp) {
            mXparser.consolePrint("Test finished, calculated = " + calculatedValue + " expected = " + expectedValue + " diff = " + MathFunctions.abs(calculatedValue - expectedValue) + " --> result ");
            if (testResult) {
                mXparser.consolePrintln("OK, time = " + testExp.getComputingTime());
            } else {
                mXparser.consolePrintln("ERROR, time = " + testExp.getComputingTime());
                consolePrintOptionsExpression(testExp);
            }
        }

        internal static void consolePrintTestSynEnd(bool receivedSyntax, bool expectedSyntax, bool testResult, Expression testExp) {
            mXparser.consolePrint("Test finished, received = " + receivedSyntax + " expected = " + expectedSyntax + " --> result ");
            if (testResult) {
                mXparser.consolePrintln("OK, time = " + testExp.getComputingTime());
            } else {
                mXparser.consolePrintln("ERROR, time = " + testExp.getComputingTime());
                consolePrintOptionsExpression(testExp);
            }
        }


        internal static void consolePrintTestApiEnd(bool testResult) {
            mXparser.consolePrint("Test finished, result = " + testResult  + " --> result ");
            if (testResult) {
                mXparser.consolePrintln("OK");
            } else {
                mXparser.consolePrintln("ERROR");
            }
        }


        internal static void consolePrintTestSynStart(int testId, String testStr) {
            consolePrintTestStart(testId, testStr, "Syn");
        }

        internal static void consolePrintTestApiStart(int testId, String testDescr) {
            consolePrintTestStart(testId, testDescr, "API");
        }

        internal static void testExprSettingsInit() {
            mXparser.setEpsilonComparison();
            mXparser.setDefaultEpsilon();
            mXparser.enableUlpRounding();
        }
        internal static void testSynSettingsInit() {
            mXparser.setDefaultOptions();
        }
        internal static void testApiSettingsInit() {
            mXparser.setDefaultOptions();
        }
    }


	/**
	 * Example of implementation
	 * FunctionExtension interface
	 * @see FunctionExtension
	 */
	internal class FunExt : FunctionExtension {
		double x;
		double y;
		public FunExt() {
			x = Double.NaN;
			y = Double.NaN;
		}
		public FunExt(double x, double y) {
			this.x = x;
			this.y = y;
		}
		public int getParametersNumber() {
			return 2;
		}
		public void setParameterValue(int parameterIndex, double parameterValue) {
			if (parameterIndex == 0) x = parameterValue;
			if (parameterIndex == 1) y = parameterValue;
		}
		public String getParameterName(int parameterIndex) {
			if (parameterIndex == 0) return "x";
			if (parameterIndex == 1) return "y";
			return "";
		}
		public double calculate() {
			return x * y;
		}
		public FunctionExtension clone() {
			return new FunExt(x, y);
		}
	}
	/**
	 * Example of implementation
	 * FunctionExtensionVariadic interface
	 * @see FunctionExtensionVariadic
	 */
	internal class FunExtVar : FunctionExtensionVariadic {
		public double calculate(params double[] parameters) {
			if (parameters == null) return Double.NaN;
			if (parameters.Length == 0) return Double.NaN;
			double result = 0;
			foreach (double x in parameters)
				result+=x;
			return result;
		}
		public FunctionExtensionVariadic clone() {
			return new FunExtVar();
		}
	}
	/**
	 * Example of implementation
	 * ArgumentExtension interface
	 * @see ArgumentExtension
	 */
	internal class PiMultArgExt : ArgumentExtension {
		private int multiple = 0;
		public double getArgumentValue() {
			multiple++;
			return  MathConstants.PI * multiple;
		}
		public ArgumentExtension clone() {
			return new PiMultArgExt();
		}
	}

}